Feature: REST client patch request

bmaxime-c opened the pull request:
Add http PATCH request on RestClient. ✅
Run dotnet format on the whole solution. ✅
Many typo fixes. ✅

devpro commented on test/AspNetCoreApiSample.IntegrationTests/Resources/TaskResourceTest.cs:
```diff
@@ -64,7 +66,8 @@ public async Task AspNetCoreApiSampleTaskResourceFullCycle_IsOk()
             };
-            await _restRunner.GetResourceByIdAsync(created.Id, expectedNotFound, HttpStatusCode.NotFound, config => config.Excluding(x => x.Extensions));
+            await _restRunner.GetResourceByIdAsync(created.Id, expectedNotFound, HttpStatusCode.NotFound,
+                config => config.Excluding(x => x.Extensions));
```
Would it be possible to remove all the added line returns?

devpro commented on samples/AspNetCoreApiSample/Utils/PropertiesChangeTracker.cs:
```diff
@@ -0,0 +1,58 @@
+    /// <summary>
+    /// A tracker that allows to keep track of any changed properties
+    /// </summary>
+    public class PropertiesChangedTracker : INotifyPropertyChanged
```
Why do we need this file?
I don't see it used in code referencing TaskPatchDto.
(BTW Utils is a too generic word)
It would be best to have the code as much simple as possible.

bmaxime-c replied:
This class allows to track which properties were changed since the instanciation or reinitialization (with AcceptChanges method).
This is useful to differentiate for example null properties because the value was never modified (so in a PATCH behavior, this value will not be modified in repository), from null value which was explicitly set (the value has to be erased in the repository).
There are some libraries which allows this tracking, but they are not free (PostSharp) or need a configuration that seems to be not compliant with hexagonal architecture (eg: TrackerDog).

devpro replied:
Interesting!
Would it be possible to add this test scenario in the integration tests? (do a patch with at least one null value then check the values returned by the get)

devpro commented on samples/AspNetCoreApiSample/Dto/TaskPatchDto.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Task Patch DTO (Data Transfer Object).
+    /// </summary>
+    public class TaskPatchDto : PropertiesChangedTracker
```
Remove PropertiesChangedTracker inheritance (not used) and use simple get set properties

devpro reviewed (changes requested):
Small but needed changes to be compliant.

devpro commented on test/AspNetCoreApiSample.IntegrationTests/Resources/TaskResourceTest.cs:
```diff
@@ -91,7 +93,8 @@ public async Task AspNetCoreApiSampleTaskResourceUpdate_WhenResourceDoesNotExist
 
         [Fact]
-        public async Task AspNetCoreApiSampleTaskResourceUpdate_WhenResourceDoesNotExistAndNewTitleIsNull_ReturnsHttpBadRequest()
+        public async Task
```
Line ending to remove :)

devpro commented on test/AspNetCoreApiSample.IntegrationTests/Resources/TaskResourceTest.cs:
```diff
@@ -106,7 +109,8 @@ public async Task AspNetCoreApiSampleTaskResourceUpdate_WhenResourceDoesNotExist
 
         [Fact]
-        public async Task AspNetCoreApiSampleTaskResourceUpdate_WhenResourceDoesNotExistAndNewTitleIsNull_ReturnsHttpLighBadRequest()
+        public async Task
```
Line ending to remove :)

devpro commented on test/AspNetCoreApiSample.IntegrationTests/Resources/TaskResourceTest.cs:
```diff
@@ -147,5 +151,47 @@ public async Task AspNetCoreApiSampleTaskResourceUpdate_WhenNullDto_ReturnsHttpB
+        }
+
+        [Fact]
+        public async Task
```
Line ending to remove :)

devpro commented:
There is an issue with secret management in the pipeline for pull request coming from forks, I'll bypass the check so the PR can be approve when comments are resolved.

bmaxime-c commented:
I removed unexpected line endings and added a new test method AspNetCoreApiSampleTaskResourcePatch_WhenModifyOneProperty_ReturnsExpectedObject which verifies that only one property is updated when using patch method.

devpro reviewed (changes requested):
@bmaxime-c One missing thing I should better document in the README, can you please update Directory.Build.props file to increase the minor version value?

This is mandatory to have a new version of the NuGet packaqes to be generated.
Once done I'll approve right away the PR and all packages will be generated.

bmaxime-c commented:
Done, the version number is now 1.3.3
